### PR TITLE
[ConvDiff] Extract system matrix and vector to scipy/numpy

### PR DESCRIPTION
**Pull request by armingeiser**
**Description**
A solver that extracts system matrix and vector as scipy/numpy without actually solving. Based on the `applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_base_solver.py`

**Changelog**
- Add stationary convection diffusion solver for system matrix and vector extraction to scipy and numpy

**Example usage**
```python
import KratosMultiphysics
from KratosMultiphysics.ConvectionDiffusionApplication.convection_diffusion_analysis import ConvectionDiffusionAnalysis

import scipy.sparse

def main():
    with open("ProjectParameters.json",'r') as parameter_file:
        parameters = KratosMultiphysics.Parameters(parameter_file.read())

    model = KratosMultiphysics.Model()
    simulation = ConvectionDiffusionAnalysis(model,parameters)
    simulation.Run()

    K = simulation._GetSolver().K  # system matrix
    p = simulation._GetSolver().p  # system vector

    matrix_file_name = "K.npz"
    print("Storing system matrix as {}. Hint: Use scipy.sparse.load_npz to read it.".format(matrix_file_name))
    scipy.sparse.save_npz(matrix_file_name, K)

    print("--Finished!--")

if __name__ == "__main__":
    main()
```

### CODE REVIEW

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_stationary_matrix_solver.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+        # example of a solution with the extracted matrix and vector
+        # ----
+        # import scipy.sparse.linalg
+        # x = scipy.sparse.linalg.spsolve(self.K, self.p)
+
+        # self._AssignSystemVector(x)
+        # ----
+
+        KratosMultiphysics.Logger.PrintInfo("::[ConvectionDiffusionStationaryMatrixSolver]:: ", "Extracted system matrix and vector.")
```
Comment (sunethwarna) [inline applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_stationary_matrix_solver.py]: just curious, is there a reason not to use `self.__class__.__name__` here? :)

File: applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_stationary_matrix_solver.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        A_csr = KratosMultiphysics.scipy_conversion_tools.to_csr(A)
+        # Convert system vector to np
+        b_np = np.array(b)
+
+        return A_csr, b_np
+
+
+    def _AssignSystemVector(self, vector):
+        """Assigns the values of the vector to the TEMPERATURE dofs """
+        print(len(vector))
```
Comment (sunethwarna) [inline applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_stationary_matrix_solver.py]: `print(len(vector))`
This is a debug thing ryt? Can be removed?

**Review by sunethwarna — DISMISSED** [review]
Minor comment only. Thanks :)

**Comment by manuelmessmer** [comment]
@armingeiser Have you check if you can call this from MainKratos.py? If I remember correct, I still had to provide the linear solver and strategy to make the workflow running.

File: applications/ConvectionDiffusionApplication/python_scripts/python_solvers_wrapper_convection_diffusion.py [hunk 3]
```diff
@@ -36,6 +36,8 @@ def CreateSolverByParameters(model, solver_settings, parallelism):
         # Steady solver
         elif (solver_type == "stationary" or solver_type == "Stationary"):
             solver_module_name = "convection_diffusion_stationary_solver"
+        elif (solver_type == "stationary_matrix"):
```
Comment (manuelmessmer) [inline applications/ConvectionDiffusionApplication/python_scripts/python_solvers_wrapper_convection_diffusion.py:39]: This is not necessary, you could also call the module by providing the folder/file_name in the .json. But depends on how you wanna have it.

**Comment by manuelmessmer** [comment]
> @armingeiser Have you check if you can call this from MainKratos.py? If I remember correct, I still had to provide the linear solver and strategy to make the workflow running.

Just saw that you derive from a customized solver and the respective functions are in there.

**Comment by manuelmessmer** [comment]
Looks good to me!

File: applications/ConvectionDiffusionApplication/python_scripts/python_solvers_wrapper_convection_diffusion.py (hunk 3 above)
Comment (armingeiser) [reply applications/ConvectionDiffusionApplication/python_scripts/python_solvers_wrapper_convection_diffusion.py:39]: how would that work?

**Comment by armingeiser** [comment]
> Just saw that you derive from a customized solver and the respective functions are in there.
Yes, since i want to reuse all the preprocessing that is happening there.

File: applications/ConvectionDiffusionApplication/python_scripts/python_solvers_wrapper_convection_diffusion.py (hunk 3 above)
Comment (manuelmessmer) [reply applications/ConvectionDiffusionApplication/python_scripts/python_solvers_wrapper_convection_diffusion.py:39]: https://github.com/KratosMultiphysics/Kratos/blob/b33095dd55ad719d7b6e04aa69dee03be552c1bf/applications/StructuralMechanicsApplication/tests/test_custom_scipy_base_solver.py#L33
Well not the folder, but the python module.

Comment (manuelmessmer) [reply applications/ConvectionDiffusionApplication/python_scripts/python_solvers_wrapper_convection_diffusion.py:39]: But probably, it is only available in the StructuralMechanicsApplication. The solver_wrapper_structrual.py checks if it finds a module with the respective name.

Comment (armingeiser) [reply applications/ConvectionDiffusionApplication/python_scripts/python_solvers_wrapper_convection_diffusion.py:39]: Interesting. This mechanism is not implemented in ConvectionDiffusionApplication however. But good to know this exists (at least for StructMech).

**Review by manuelmessmer — APPROVED** [review]

**Comment by rubenzorrilla** [comment]
Why is this required? Can't the MatrixMarket tools already available be used to output the A and b?

**Comment by armingeiser** [comment]
> Why is this required? Can't the MatrixMarket tools already available be used to output the A and b?

That allows wrinting the matrix to disk, but not to get access to the matrix in scipy/numpy in python directly (of course one could write to disk and then read in python).
Also with the existing tool you will always have to do the system solution to get the output of the matrix, which is technically not required.

**Comment by rubenzorrilla** [comment]
> > Why is this required? Can't the MatrixMarket tools already available be used to output the A and b?
> 
> That allows writing the matrix to disk, but not to get access to the matrix in scipy/numpy in python directly (of course one could write to disk and then read in python).
> Also with the existing tool you will always have to do the system solution to get the output of the matrix, which is technically not required.

OK. Thanks for the explanation @armingeiser . If it eases your research I'm fine with adding it.

**Pull request merged**